pkg/sink/file/writer.go

```diff
@@ -0,0 +1,163 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package file
+
+import (
+	"bufio"
+	"io"
+	"sync"
+	"time"
+)
+
+const (
+	// _defaultBufferSize specifies the default size used by Buffer.
+	_defaultBufferSize = 256 * 1024 // 256 kB
+
+	// _defaultFlushInterval specifies the default flush interval for
+	// Buffer.
+	_defaultFlushInterval = 30 * time.Second
+)
+
+type Writer struct {
+	// This field is required.
+	W io.Writer
+
+	// Size specifies the maximum amount of data the writer will buffered
+	// before flushing.
+	//
+	// Defaults to 256 kB if unspecified.
+	Size int
+
+	// FlushInterval specifies how often the writer should flush data if
+	// there have been no writes.
+	//
+	// Defaults to 30 seconds if unspecified.
+	FlushInterval time.Duration
+
+	// unexported fields for state
+	mu          sync.Mutex
+	initialized bool // whether initialize() has run
+	stopped     bool // whether Stop() has run
+	writer      *bufio.Writer
+	ticker      *time.Ticker
+	stop        chan struct{} // closed when flushLoop should stop
+	done        chan struct{} // closed when flushLoop has stopped
+}
+
+func (w *Writer) initialize() {
+	size := w.Size
+	if size == 0 {
+		size = _defaultBufferSize
+	}
+
+	flushInterval := w.FlushInterval
+	if flushInterval == 0 {
+		flushInterval = _defaultFlushInterval
+	}
+
+	w.ticker = time.NewTicker(flushInterval)
+	w.writer = bufio.NewWriterSize(w.W, size)
+	w.stop = make(chan struct{})
+	w.done = make(chan struct{})
+	w.initialized = true
+	go w.flushLoop()
+}
+
+// Write writes log data into buffer syncer directly, multiple Write calls will be batched,
+// and log data will be flushed to disk when the buffer is full or periodically.
+func (w *Writer) Write(bs []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if !w.initialized {
+		w.initialize()
+	}
+
+	// To avoid partial writes from being flushed, we manually flush the existing buffer if:
+	// * The current write doesn't fit into the buffer fully, and
+	// * The buffer is not empty (since bufio will not split large writes when the buffer is empty)
+	if len(bs) > w.writer.Available() && w.writer.Buffered() > 0 {
+		if err := w.writer.Flush(); err != nil {
+			return 0, err
+		}
+	}
+
+	return w.writer.Write(bs)
+}
+
+// Sync flushes buffered log data into disk directly.
+func (w *Writer) Sync() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.initialized {
+		return w.writer.Flush()
+	}
+
+	return nil
+}
+
+// flushLoop flushes the buffer at the configured interval until Stop is
+// called.
+func (w *Writer) flushLoop() {
+	defer close(w.done)
+
+	for {
+		select {
+		case <-w.ticker.C:
+			// we just simply ignore error here
+			// because the underlying bufio writer stores any errors
+			// and we return any error from Sync() as part of the close
+			_ = w.Sync()
+		case <-w.stop:
+			return
+		}
+	}
+}
```

Comment:
can flushloop be processed globally with one goroutine?

Reply:
If use one goroutine, it may need a writer manager. Multiwriter can do it, but the writer is incomplete, so we plan to not change it first. If there is a problem with performance, we can consider how to optimize it together.

Reply:
I turned off flushloop switch in writer and used one goroutine in multiwriter to handle flushing of multiple writers